server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalRareTerms.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.search.aggregations.bucket.terms;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.util.ExactBloomFilter;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.aggregations.Aggregations;
+import org.elasticsearch.search.aggregations.BucketOrder;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.InternalAggregations;
+import org.elasticsearch.search.aggregations.InternalMultiBucketAggregation;
+import org.elasticsearch.search.aggregations.InternalOrder;
+import org.elasticsearch.search.aggregations.KeyComparable;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public abstract class InternalRareTerms<A extends InternalRareTerms<A, B>, B extends InternalRareTerms.Bucket<B>>
+    extends InternalMultiBucketAggregation<A, B> implements RareTerms {
+
+    public abstract static class Bucket<B extends Bucket<B>> extends InternalMultiBucketAggregation.InternalBucket
+        implements RareTerms.Bucket, KeyComparable<B> {
+        /**
+         * Reads a bucket. Should be a constructor reference.
+         */
+        @FunctionalInterface
+        public interface Reader<B extends Bucket<B>> {
+            B read(StreamInput in, DocValueFormat format) throws IOException;
+        }
+
+        long bucketOrd;
```

Comment:
@polyfractal could you comment here, I don't see how this is used and it is not part of equal or hashCode?

Reply:
@iverase  Right thanks, forgot to comment on this.  So this is a slightly strange idiom that the terms agg uses, and rare_terms just inherited the same behavior.

The `bucketOrd` on each bucket is a temporary "scratch" space for noting down the bucket ordinal at runtime, which is used to run deferred aggs after collecting the relevant buckets.  But it's not serialized between nodes, and not part of the equals/hashcode/etc.

E.g. when we are building the sketch, we skip buckets that are over-threshold and [re-assign ordinals of candidate buckets](https://github.com/elastic/elasticsearch/pull/35718/files#diff-29eed13231a9e57022105c0d193c398cR120).  The new ordinals are then used for [deferred aggregations](https://github.com/elastic/elasticsearch/pull/35718/files#diff-29eed13231a9e57022105c0d193c398cR149) after we merge out the empty buckets.

Regular terms have a [similar `bucketOrd](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalTerms.java#L60) and uses it as a temporary scratch space to note down ordinals when running the [top-n priority queue](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongTermsAggregator.java#L144)

It's not a great pattern and it'd be nice if we could remove it, but I think we should make the change to all the terms/sigterms/rareterms at the same time since it's a pattern used across all of them.

Reply:
> I think we should make the change to all the terms/sigterms/rareterms at the same time since it's a pattern used across all of them.

I agree and if we cannot remove it at least rename it to something that suggest is a scratch value :)